phat/contracts/lego/js/src/index.ts

```diff
@@ -1,190 +1,279 @@
-  const args = action.config
-  const output = pink.invokeContract({
-    ...args,
-    input
-  })
-  return output
-}
+  function actionCall(
+    action: ActionCall,
+    input: any,
+    typeRegistry: TypeRegistry
+  ): Uint8Array {
+    const args = action.config;
+    const codec = action.config.codec;
+    if (codec?.inputs !== undefined) {
+      input = encode(input, codec.inputs, typeRegistry);
+      if (debugWorkflow) {
+        console.log(`encoded call input: ${hex(input)}`);
+      }
+    }
+    let output = $pink.invokeContract({
+      ...args,
+      input,
+    });
+    if (debugWorkflow) {
+      console.log(`call output: ${repr(output)}`);
+    }
+    if (codec?.output !== undefined) {
+      output = decode(output, codec.output, typeRegistry);
+    }
+    return output;
+  }
 
-function actionEval(action: ActionEval, input: any, context: any): any {
-  const script = action.config
-  if (typeof script !== 'string') {
-    throw new Error('Trying to eval non-string')
-  }
-
-  // pub fn build_transaction(
-  //   &self,
-  //   to: String,
-  //   abi: Vec<u8>,
-  //   func: String,
-  //   params: Vec<Vec<u8>>,
-  // )
-  type Tuple = [string, number[], string, (number[])[]];
-  const encodeBuildTx = scaleCore.createTupleEncoder<Tuple>([scaleCore.encodeStr, scaleCore.createVecEncoder(scaleCore.encodeU8), scaleCore.encodeStr, scaleCore.createVecEncoder(scaleCore.createVecEncoder(scaleCore.encodeU8))]);
-
-  const decodeResultVecU8 = scaleCore.createResultDecoder<number[], any>(
-    scaleCore.createVecDecoder(scaleCore.decodeU8),
-    scaleCore.createEnumDecoder({
-      0: 'BadOrigin',
-      1: 'NotConfigured',
-      2: 'BadAbi',
-      3: 'BadParams',
-      4: 'BadToAddress',
-      5: 'BadTransaction',
-      6: 'FailedToSendTransaction'
-    })
-  );
-
-  const scale = {
-    encode: scaleCore.WalkerImpl.encode,
-    decode: scaleCore.WalkerImpl.decode,
-    encodeU128: scaleCore.encodeU128,
-    encodeU64: scaleCore.encodeU64,
-    encodeU32: scaleCore.encodeU32,
-    encodeU16: scaleCore.encodeU16,
-    encodeUint8Vec: scaleCore.encodeUint8Vec,
-    encodeStr: scaleCore.encodeStr,
-    encodeBuildTx: encodeBuildTx,
-    encodeVecU8: scaleCore.createVecEncoder(scaleCore.encodeU8),
-    decodeResultVecU8: decodeResultVecU8,
-    createStructEncoder: scaleCore.createStructEncoder,
-    createEnumEncoder: scaleCore.createEnumEncoder
+  function actionEval(action: ActionEval, input: any, context: any): any {
+    const script = action.config;
+    if (typeof script !== "string") {
+      throw new Error("Trying to eval non-string");
+    }
+    return doEval(script, input, context);
   }
 
-  function numToUint8Array32(num: number) {
-    let arr = new Uint8Array(32);
+  function actionScale(
+    action: ActionScale,
+    input: any,
+    typeRegistry: TypeRegistry
+  ): any {
+    const { subcmd, type } = action.config;
+    const codec = $.codec(type, typeRegistry);
```

Comment:
A bit heavy and it was cached in previous implementation. But we can not cache it because we encontered OOM in some case and I was struggling to reduce the memory usage, so removed the cache of this.